restore/data.go

```diff
@@ -215,34 +215,49 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
 		gplog.Verbose("Initializing pipes and gpbackup_helper on segments for %srestore", msg)
 		utils.VerifyHelperVersionOnSegments(version, globalCluster)
 
-		// During a larger-to-smaller restore, we need to do multiple passes of
-		// data loading so we assign the batches here.
-		oidList := make([]string, 0)
-		for _, entry := range dataEntries {
-			if entry.IsReplicated {
-				oidList = append(oidList, fmt.Sprintf("%d,0", entry.Oid))
-				continue
-			}
+		var maxHelpers int
+		if backupConfig.SingleDataFile {
+			maxHelpers = 1
+		} else if connectionPool.NumConns < totalTables {
+			maxHelpers = connectionPool.NumConns
+		} else {
+			maxHelpers = totalTables
+		}
+
+		for helperIdx := 0; helperIdx < maxHelpers; helperIdx++ {
+			// During a larger-to-smaller restore, we need to do multiple passes of
+			// data loading so we assign the batches here.
+			oidList := make([]string, 0)
+			for entryIdx, entry := range dataEntries {
+				if (entryIdx % maxHelpers) != helperIdx {
+					continue
```

Comment:
I think we are wasting here too many operations when filtering out entries. On a real world backup that could be millions of skipped entries for each helperIdx, which we could have avoided. What about the change below?
```diff
diff --git a/restore/data.go b/restore/data.go
index 65e3e3e2..73c5a570 100644
--- a/restore/data.go
+++ b/restore/data.go
@@ -228,18 +228,14 @@ func restoreDataFromTimestamp(fpInfo filepath.FilePathInfo, dataEntries []toc.Co
                        // During a larger-to-smaller restore, we need to do multiple passes of
                        // data loading so we assign the batches here.
                        oidList := make([]string, 0)
-                       for entryIdx, entry := range dataEntries {
-                               if (entryIdx % maxHelpers) != helperIdx {
-                                       continue
-                               }
-
-                               if entry.IsReplicated {
-                                       oidList = append(oidList, fmt.Sprintf("%d,0", entry.Oid))
+                       for entryIdx := helperIdx; entryIdx < len(dataEntries); entryIdx += maxHelpers {
+                               if dataEntries[entryIdx].IsReplicated {
+                                       oidList = append(oidList, fmt.Sprintf("%d,0", dataEntries[entryIdx].Oid))
                                        continue
                                }
 
                                for b := 0; b < batches; b++ {
-                                       oidList = append(oidList, fmt.Sprintf("%d,%d", entry.Oid, b))
+                                       oidList = append(oidList, fmt.Sprintf("%d,%d", dataEntries[entryIdx].Oid, b))
                                }
                        }

```

Reply:
applied